File: go/pkg/hey/entries.go
package hey

import (
	"context"
	"time"

	"github.com/basecamp/hey-sdk/go/pkg/generated"
)

// EntriesService handles draft and reply operations.
type EntriesService struct {
	client *Client
}

// NewEntriesService creates a new EntriesService.
func NewEntriesService(client *Client) *EntriesService {
	return &EntriesService{client: client}
}

// ListDrafts returns all draft messages.
func (s *EntriesService) ListDrafts(ctx context.Context, params *generated.ListDraftsParams) (result *generated.ListDraftsResponseContent, err error) {
	op := OperationInfo{
		Service: "Entries", Operation: "ListDrafts",
		ResourceType: "draft", IsMutation: false,
	}
	if gater, ok := s.client.hooks.(GatingHooks); ok {
		if ctx, err = gater.OnOperationGate(ctx, op); err != nil {
			return
		}
	}
	start := time.Now()
	ctx = s.client.hooks.OnOperationStart(ctx, op)
	defer func() { s.client.hooks.OnOperationEnd(ctx, op, err, time.Since(start)) }()

	s.client.initGeneratedClient()
	resp, err := s.client.gen.ListDraftsWithResponse(ctx, params)
	if err != nil {
		return nil, err
	}
	if err = CheckResponse(resp.HTTPResponse); err != nil {
		return nil, err
	}
	return resp.JSON200, nil
}

// CreateReply replies to an entry (POST /entries/{entryId}/replies.json) and delivers it.
//
// ActingSenderID is the identity the reply goes out as. HEY resolves a reply's sender
// from the thread — on a shared or alternate address (a HEY for Work support address,
// an extension, a forwarded external account) that is not the account default — and
// hands it back as NewReply's Sender. Pass that prefill's Sender.Id; zero falls back
// to the account's default sender, which on such a thread is the wrong identity.
//
// Subject is the reply's own subject line — HEY does not derive one, so pass the
// prefilled "Re: …" that NewReply hands back. Empty leaves it off the wire.
//
// Recipients are required. HEY does not reply-all on the caller's behalf: a reply
// posted without entry.addressed is saved as a draft (the server answers with a
// redirect to the thread with the draft expanded) rather than delivered. Callers
// resolve the thread's recipients first — hey-cli reads them from the topic page.
func (s *EntriesService) CreateReply(ctx context.Context, entryID, actingSenderID int64, subject, content string, to, cc, bcc []string) (err error) {
	if len(to)+len(cc)+len(bcc) == 0 {
		return ErrUsage("a reply needs at least one recipient (to, cc or bcc); HEY saves an unaddressed reply as a draft")
	}
	op := OperationInfo{
		Service: "Entries", Operation: "CreateReply",
		ResourceType: "reply", IsMutation: true, ResourceID: entryID,
	}
	if gater, ok := s.client.hooks.(GatingHooks); ok {
		if ctx, err = gater.OnOperationGate(ctx, op); err != nil {
			return
		}
	}
	start := time.Now()
	ctx = s.client.hooks.OnOperationStart(ctx, op)
	defer func() { s.client.hooks.OnOperationEnd(ctx, op, err, time.Since(start)) }()

	senderID, err := s.resolveActingSenderID(ctx, actingSenderID)
	if err != nil {
		return err
	}

	body := generated.CreateReplyRequestContent{
		ActingSenderId: senderID,
		Message:        generated.ReplyMessagePayload{Subject: subject, Content: content},
		Entry:          entryPayload(to, cc, bcc),
	}

	resp, err := s.client.genClient().CreateReplyWithResponse(ctx, entryID, body)
	if err != nil {
		return err
	}
	return CheckResponse(resp.HTTPResponse)
}

// resolveActingSenderID answers the acting sender a reply is sent as: the caller's
// choice when one was made, the account's default sender otherwise.
func (s *EntriesService) resolveActingSenderID(ctx context.Context, actingSenderID int64) (int64, error) {
	if actingSenderID > 0 {
		return actingSenderID, nil
	}
	return s.client.DefaultSenderID(ctx)
}

// MarkSpam marks an entry as spam. The server denies the sender outright when every thread
// from them is already spam.
func (s *EntriesService) MarkSpam(ctx context.Context, entryID int64) error {
	op := OperationInfo{
		Service: "Entries", Operation: "MarkEntrySpam",
		ResourceType: "entry", IsMutation: true, ResourceID: entryID,
	}

	return s.client.instrument(ctx, op, func(ctx context.Context) error {
		resp, err := s.client.genClient().MarkEntrySpamWithResponse(ctx, entryID)
		if err != nil {
			return err
		}
		return CheckResponse(resp.HTTPResponse)
	})
}

// DraftPage is one page of drafts along with the cursor for the page after it. The
// index pages by geared_pagination's opaque cursor out of the Link header — a page
// number is answered with the first page forever — so NextPage is the only way to walk
// it; empty means the last page.
type DraftPage struct {
	Drafts   []generated.DraftMessage
	NextPage string
}

// ListDraftsPage answers the same drafts as ListDrafts along with the cursor for the
// page after them.
func (s *EntriesService) ListDraftsPage(ctx context.Context, page string) (result *DraftPage, err error) {
	op := OperationInfo{
		Service: "Entries", Operation: "ListDrafts",
		ResourceType: "draft", IsMutation: false,
	}

	err = s.client.instrument(ctx, op, func(ctx context.Context) error {
		params := &generated.ListDraftsParams{}
		if page != "" {
			params.Page = &page
		}

		resp, rerr := s.client.genClient().ListDraftsWithResponse(ctx, params)
		if rerr != nil {
			return rerr
		}
		if cerr := CheckResponse(resp.HTTPResponse); cerr != nil {
			return cerr
		}
		result = &DraftPage{}
		if resp.JSON200 != nil {
			result.Drafts = *resp.JSON200
		}
		if resp.HTTPResponse != nil {
			result.NextPage = gearedPageFromLink(resp.HTTPResponse.Header.Get("Link"))
		}
		return nil
	})
	return result, err
}

// DeleteDraft trashes a draft by its entry id, as ListDrafts reports it.
func (s *EntriesService) DeleteDraft(ctx context.Context, entryID int64) error {
	op := OperationInfo{
		Service: "Entries", Operation: "DeleteDraft",
		ResourceType: "draft", IsMutation: true, ResourceID: entryID,
	}

	return s.client.instrument(ctx, op, func(ctx context.Context) error {
		resp, err := s.client.genClient().DeleteDraftWithResponse(ctx, entryID)
		if err != nil {
			return err
		}
		return CheckResponse(resp.HTTPResponse)
	})
}

// CreateReplyDraft saves a reply to an entry as a draft instead of delivering it, and
// answers the draft's entry id. Unlike CreateReply it needs no recipients — HEY keeps
// whatever the draft carries.
//
// ActingSenderID works as it does on CreateReply: pass the prefill's Sender.Id from
// NewReply so the draft is saved under the identity HEY resolved for the thread; zero
// falls back to the account's default sender.
//
// Subject matters here: HEY does not derive one, and a reply draft saved without it
// shows as "No subject" in Drafts. Pass the prefilled "Re: …" from NewReply. Empty
// leaves it off the wire.
func (s *EntriesService) CreateReplyDraft(ctx context.Context, entryID, actingSenderID int64, subject, content string, to, cc, bcc []string) (draftEntryID int64, err error) {
	op := OperationInfo{
		Service: "Entries", Operation: "CreateReply",
		ResourceType: "draft", IsMutation: true, ResourceID: entryID,
	}

	err = s.client.instrument(ctx, op, func(ctx context.Context) error {
		senderID, serr := s.resolveActingSenderID(ctx, actingSenderID)
		if serr != nil {
			return serr
		}

		entry := &generated.MessageEntryPayload{
			Status:    entryStatusDrafted,
			Addressed: &generated.MessageAddressed{Directly: to, Copied: cc, Blindcopied: bcc},
		}
		body := generated.CreateReplyRequestContent{
			ActingSenderId: senderID,
			Message:        generated.ReplyMessagePayload{Subject: subject, Content: content},
			Entry:          entry,
		}

		resp, rerr := s.client.genClient().CreateReplyWithResponse(ctx, entryID, body)
		if rerr != nil {
			return rerr
		}
		if cerr := CheckResponse(resp.HTTPResponse); cerr != nil {
			return cerr
		}
		draftEntryID, rerr = draftEntryIDFromLocation(resp.HTTPResponse)
		return rerr
	})
	return draftEntryID, err
}

// NewReply returns a prefilled reply to an entry: the quoted body; in Addressed, the
// recipients a reply goes to as HEY computes them — the entry's sender moved onto the
// To line and the acting user's own addresses, aliases and catch-alls excluded; and in
// Sender, the identity HEY resolved for the reply from the thread's own to and from
// addresses, present only when it differs from the acting user. This is the recipient
// list and acting sender CreateReply should be handed.
func (s *EntriesService) NewReply(ctx context.Context, entryID int64) (result *generated.MessageDraft, err error) {
	op := OperationInfo{
		Service: "Entries", Operation: "NewEntryReply",
		ResourceType: "entry", IsMutation: false, ResourceID: entryID,
	}

	err = s.client.instrument(ctx, op, func(ctx context.Context) error {
		resp, rerr := s.client.genClient().NewEntryReplyWithResponse(ctx, entryID)
		if rerr != nil {
			return rerr
		}
		if cerr := CheckResponse(resp.HTTPResponse); cerr != nil {
			return cerr
		}
		result = resp.JSON200
		return nil
	})
	return result, err
}

// NewForward returns a prefilled forward of an entry: the "Fwd:" subject, the quoted body and
// blank recipients. Fill in the recipients and send it with MessagesService.Create.
func (s *EntriesService) NewForward(ctx context.Context, entryID int64) (result *generated.MessageDraft, err error) {
	op := OperationInfo{
		Service: "Entries", Operation: "NewEntryForward",
		ResourceType: "entry", IsMutation: false, ResourceID: entryID,
	}

	err = s.client.instrument(ctx, op, func(ctx context.Context) error {
		resp, rerr := s.client.genClient().NewEntryForwardWithResponse(ctx, entryID)
		if rerr != nil {
			return rerr
		}
		if cerr := CheckResponse(resp.HTTPResponse); cerr != nil {
			return cerr
		}
		result = resp.JSON200
		return nil
	})
	return result, err
}

File: rust/hey-sdk/src/services/entries.rs
//! Replying to an entry, delivered or kept as a draft.

use crate::error::Error;
use crate::generated::routes;
use crate::generated::types::{CreateReplyRequestContent, ReplyMessagePayload};
use crate::services::messages::{
    delivered_entry, drafted_entry, entry_id_from_location, has_recipients,
};

pub use crate::generated::services::entries::*;

/// A reply, as the reply prefill (`Entries::new_reply`) hands its parts over.
#[derive(Debug, Clone, Default, PartialEq)]
pub struct ReplyContent {
    /// The identity the reply goes out as. HEY resolves a reply's sender from the thread
    /// — on a shared or alternate address that is not the account default — and hands it
    /// back as the prefill's sender; pass that id. Zero falls back to the client's
    /// default sender, which on such a thread is the wrong identity.
    pub acting_sender_id: i64,
    /// HEY derives no subject for a reply, so pass the prefill's "Re: …" — a draft saved
    /// without one reads "No subject". Empty leaves it off the wire.
    pub subject: String,
    /// The reply body alone: HEY appends the quoted original at delivery, so the
    /// prefill's quoted content must not be echoed back.
    pub content: String,
    pub to: Vec<String>,
    pub cc: Vec<String>,
    pub bcc: Vec<String>,
}

impl<'a> Entries<'a> {
    /// Delivers a reply to an entry. HEY does not reply-all on the caller's behalf, and
    /// saves an unaddressed reply as a draft rather than delivering it, so the thread's
    /// recipients — the prefill's — are required.
    pub async fn reply(&self, entry_id: i64, reply: &ReplyContent) -> Result<(), Error> {
        if !has_recipients(&reply.to, &reply.cc, &reply.bcc) {
            return Err(Error::usage(
                "a reply needs at least one recipient (to, cc or bcc); HEY saves an unaddressed reply as a draft",
            ));
        }

        let body = CreateReplyRequestContent {
            acting_sender_id: self.sender_for(reply).await?,
            message: reply_payload(reply),
            entry: Some(delivered_entry(&reply.to, &reply.cc, &reply.bcc)),
        };
        let mut operation = self.client().operation(&routes::CREATE_REPLY, &[&entry_id]);
        operation.json(&body)?;
        self.client().send_unit(operation).await
    }

    /// Saves a reply as a draft instead of delivering it, and answers the draft's entry
    /// id. Unlike [`Entries::reply`] it needs no recipients; whatever it carries is kept
    /// for the send.
    pub async fn reply_draft(&self, entry_id: i64, reply: &ReplyContent) -> Result<i64, Error> {
        let body = CreateReplyRequestContent {
            acting_sender_id: self.sender_for(reply).await?,
            message: reply_payload(reply),
            entry: Some(drafted_entry(&reply.to, &reply.cc, &reply.bcc)),
        };
        let mut operation = self.client().operation(&routes::CREATE_REPLY, &[&entry_id]);
        operation.json(&body)?;
        let response = self.client().execute(operation).await?;
        entry_id_from_location(&response)
    }

    async fn sender_for(&self, reply: &ReplyContent) -> Result<i64, Error> {
        if reply.acting_sender_id == 0 {
            self.client().default_sender_id().await
        } else {
            Ok(reply.acting_sender_id)
        }
    }
}

fn reply_payload(reply: &ReplyContent) -> ReplyMessagePayload {
    let subject = if reply.subject.is_empty() {
        None
    } else {
        Some(reply.subject.clone())
    };
    ReplyMessagePayload {
        subject,
        content: reply.content.clone(),
    }
}

#[cfg(test)]
mod tests {
    use serde_json::{Value, json};
    use wiremock::matchers::{method, path};
    use wiremock::{Mock, MockServer, ResponseTemplate};

    use super::*;
    use crate::auth::StaticTokenProvider;
    use crate::client::Client;
    use crate::config::Config;
    use crate::error::ErrorCode;

    #[tokio::test]
    async fn reply_sends_the_chosen_acting_sender_untouched() {
        let server = MockServer::start().await;
        Mock::given(method("POST"))
            .and(path("/entries/456/replies.json"))
            .respond_with(ResponseTemplate::new(200).set_body_json(json!({ "notice": "sent" })))
            .mount(&server)
            .await;

        client(&server)
            .entries()
            .reply(456, &prefilled_reply())
            .await
            .unwrap();

        let body = sent_json(&server).await;
        assert_eq!(body["acting_sender_id"], 314);
        assert_eq!(
            body["message"],
            json!({ "subject": "Re: From the support address", "content": "Reply text" })
        );
        assert_eq!(
            body["entry"],
            json!({ "addressed": { "directly": ["someone@example.com"] } })
        );
    }

    #[tokio::test]
    async fn reply_refuses_a_reply_addressed_to_nobody() {
        let server = MockServer::start().await;
        let reply = ReplyContent {
            to: Vec::new(),
            ..prefilled_reply()
        };

        let error = client(&server)
            .entries()
            .reply(456, &reply)
            .await
            .unwrap_err();

        assert_eq!(error.code(), ErrorCode::Usage);
        assert!(server.received_requests().await.unwrap().is_empty());
    }

    #[tokio::test]
    async fn reply_draft_saves_it_drafted_and_answers_the_entry_id() {
        let server = MockServer::start().await;
        Mock::given(method("POST"))
            .and(path("/entries/456/replies.json"))
            .respond_with(
                ResponseTemplate::new(204)
                    .insert_header("Location", "https://app.hey.com/messages/777"),
            )
            .mount(&server)
            .await;
        let reply = ReplyContent {
            to: Vec::new(),
            content: "Draft text".to_string(),
            ..prefilled_reply()
        };

        let entry_id = client(&server)
            .entries()
            .reply_draft(456, &reply)
            .await
            .unwrap();

        assert_eq!(entry_id, 777);
        let body = sent_json(&server).await;
        assert_eq!(body["acting_sender_id"], 314);
        assert_eq!(body["message"]["subject"], "Re: From the support address");
        assert_eq!(
            body["entry"],
            json!({
                "addressed": { "directly": [], "copied": [], "blindcopied": [] },
                "status": "drafted"
            })
        );
    }

    #[tokio::test]
    async fn a_reply_without_a_subject_leaves_it_off_the_wire() {
        let server = MockServer::start().await;
        Mock::given(method("POST"))
            .and(path("/entries/456/replies.json"))
            .respond_with(ResponseTemplate::new(200).set_body_json(json!({ "notice": "sent" })))
            .mount(&server)
            .await;
        let reply = ReplyContent {
            subject: String::new(),
            ..prefilled_reply()
        };

        client(&server).entries().reply(456, &reply).await.unwrap();

        let body = sent_json(&server).await;
        assert_eq!(body["message"], json!({ "content": "Reply text" }));
    }

    fn prefilled_reply() -> ReplyContent {
        ReplyContent {
            acting_sender_id: 314,
            subject: "Re: From the support address".to_string(),
            content: "Reply text".to_string(),
            to: vec!["someone@example.com".to_string()],
            ..ReplyContent::default()
        }
    }

    fn client(server: &MockServer) -> Client {
        Client::builder(Config::default().with_base_url(server.uri()))
            .token_provider(StaticTokenProvider::new("t"))
            .max_retries(0)
            .build()
            .unwrap()
    }

    async fn sent_json(server: &MockServer) -> Value {
        let requests = server.received_requests().await.unwrap();
        assert_eq!(requests.len(), 1);
        serde_json::from_slice(&requests[0].body).unwrap()
    }
}
